File: [2] Optimize code/process.py
import os
import argparse
import time
import string


def load_dict(path_dict: str):
    """
    @arguments:
        - path_dict (string): the path of dictionary/syllabus file.
    @return values:
        - A list of word in dictionary file 
    """

    with open(path_dict, 'r', encoding="utf8") as f:
        content = f.read()
        word_list = content.splitlines()
        return word_list
    return []


def load_test(path_test:str):
    """
    @arguments:
        - path_test (string): the path of test document containt some of paragraph we need check.
    @return values:
        - the dictionary contain all word in test document and the frequency of them. 
    """
    with open(path_test, 'r', encoding="utf8") as f:
        content = f.read()
        word_list = list(map(str, content.split(" ")))
    
        # make a table for preprocess data
        table = str.maketrans('', '', string.punctuation)

        # Count the frequency of every word in the test file.
        dict_count = dict()
        
        for word in word_list:
            # preprocess data
            word = word.translate(table).upper()
            if " " not in word and "\n" not in word:
                if word in dict_count:
                    # increase frequency of word
                    dict_count[word] += 1
                else:
                    # add the word to dictionary
                    dict_count[word] = 1
        return dict_count

    return None

def display_test(document: dict):
    for item in document:
        print(item, end="   ")
        print(document[item])
    
def display_dict(dictionary: list):
    for item in dictionary:
        print(item)
    
    
def process(dictionary: list, document: dict):
    """
    @parameter:
        dictionary:
        documents:
    @return values:
        - A dictionary that the key value is the word in the dictionary and the value is the frequency of word appear
        in the documents.

    """
    # Solve the question a: count the number of word in the dictionary exist in documents.
    count_appear = 0
    for word in dictionary:
        if word in document:
            count_appear += 1

    # Solve the question b: calculate the frequency.
    frequency = count_appear / len(dictionary) * 100

    return count_appear, frequency


def main(args):
    path_dictionary = args.dict_path
    path_test_file = args.test_path
    dictionary = load_dict(path_dict=path_dictionary)
    document = load_test(path_test=path_test_file)

    # print("-------------------------The dictionary---------------------------")
    # display_dict(dictionary=dictionary)
    # print("-------------------------The test document------------------------")
    # display_test(document=document)

    # Solve problem
    count_appear, frequency = process(dictionary=dictionary, document=document)
    print(f"The number of word in dictionary appear in the test document is {count_appear}")
    print(f"The percentage is {frequency} %")
    


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Count the frequency of the words in dictionary appear in test file.')
    parser.add_argument("--dict_path", help="The path of dictionary", default="syllables.txt")
    parser.add_argument("--test_path", help="The path of test file.", default="test.txt")
    args = parser.parse_args()
    START_TIME = time.time()
    main(args)
    print("The time execution is  %s seconds ---" % (time.time() - START_TIME))
  

File: [2] Optimize code/process.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <set>
#include <vector>
#include <iomanip>
#include <string>
#include <string.h>
#include <ctype.h>
#include <algorithm>


using namespace std;

void preprocess(string&);
map<string, int> load_test(const char[]);
vector<string> load_dict(const char[]);
void display_dict(vector<string>&);
void display_test(map<string, int>&);
void toupper(string&);

set<string> SKIP_WORDS = {"~", "!", "@", "#", "$", "%", "^", "&", "*", "(", ")", "{", "}", "+", "-", "=", ".", ",", " "};

int main(int argc, char *argv[])
{
	const char test_path[] = "test.txt";
	const char dict_path[] = "syllables.txt";
	
	
	vector<string> dict = load_dict(dict_path);
	map<string, int> test = load_test(test_path);

	display_test(test);
	
	
	return 0;
}


void preprocess(string& str)
{
	int n = str.length()-1;
	int i = n;
	string temp;
	while(i >= 1){		
		temp = str[i-1] + "";
		if(SKIP_WORDS.find(temp) == SKIP_WORDS.end())
			break;
		--i;
	}
	str.erase(i, n-i);
}

map<string, int> load_test(const char path[])
{
	/*Open file*/
	ifstream in;
	in.open(path, ios::out);
	map<string, int> content;
	string word;
	while(in)
	{
		string temp;
		in >> temp;
		word = temp;
		// Preprocess word
		// preprocess(character);
		if(word != "")
		{
			if (content.find(word) == content.end())
			{
				toupper(word);
				content[word] = 1;
			}
			else
			{
				content[word]++;
			}
		}
	}
	in.close();
	return content;
}

vector<string> load_dict(const char path[]){
	/*Open file*/
	ifstream in;
	in.open(path, ios::out);
	vector<string> dict;
	string character;
	while(in)
	{
		in >> character;
		dict.push_back(character);
	}
	in.close(); // close file
	return dict;
}

void display_dict(vector<string>& dict){
	vector<string>::iterator it;
	int id = 1;
	for(it = dict.begin(); it != dict.end(); ++it)
	{
		cout << left << setw(6) << id;
		cout << *it << endl;
		++id;
	}
}


void display_test(map<string, int>& test){
	map<string, int>::iterator imap;
	for(imap = test.begin(); imap != test.end(); ++imap)
	{
		cout << imap->first << ": " << imap->second << endl;
	}
}

void toupper(string&str)
{
	std::for_each(str.begin(), str.end(), [](char & c){
		c = ::toupper(c);
	});
}
